source/fem-dem/gls_sharp_navier_stokes.cc

```diff
@@ -2901,6 +2908,99 @@ GLSSharpNavierStokesSolver<dim>::read_checkpoint()
+          particles[p_i].inertia[1][1] = restart_data["inertia"][p_i];
+          particles[p_i].inertia[2][2] = restart_data["inertia"][p_i];
+
+          particles[p_i].pressure_location[0] = restart_data["pressure_x"][p_i];
+          particles[p_i].pressure_location[1] = restart_data["pressure_y"][p_i];
+          particles[p_i].youngs_modulus = restart_data["youngs_modulus"][p_i];
+          particles[p_i].restitution_coefficient =
+            restart_data["restitution_coefficient"][p_i];
+          particles[p_i].friction_coefficient =
+            restart_data["friction_coefficient"][p_i];
+          particles[p_i].poisson_ratio = restart_data["poisson_ratio"][p_i];
+          particles[p_i].rolling_friction_coefficient =
+            restart_data["rolling_friction_coefficient"][p_i];
+          particles[p_i].initialise_last();
+        }
+    }
+  if (dim == 3)
+    {
+      // Loop over each line of the file and filling the particles properties.
+      for (unsigned int p_i = 0; p_i < particles.size(); ++p_i)
+        {
+          particles[p_i].initialise_all();
+          particles[p_i].position[0] = restart_data["p_x"][p_i];
+          particles[p_i].position[1] = restart_data["p_y"][p_i];
+          particles[p_i].position[2] = restart_data["p_z"][p_i];
+          particles[p_i].velocity[0] = restart_data["v_x"][p_i];
+          particles[p_i].velocity[1] = restart_data["v_y"][p_i];
+          particles[p_i].velocity[2] = restart_data["v_z"][p_i];
+
+          particles[p_i].radius = restart_data["radius"][p_i];
+          particles[p_i].mass   = 4.0 / 3.0 * PI * particles[p_i].radius *
+                                particles[p_i].radius * particles[p_i].radius *
+                                restart_data["density"][p_i];
+
+          particles[p_i].omega[0] = restart_data["omega_x"][p_i];
+          particles[p_i].omega[1] = restart_data["omega_y"][p_i];
+          particles[p_i].omega[2] = restart_data["omega_z"][p_i];
+
+          particles[p_i].inertia[0][0] = restart_data["inertia"][p_i];
+          particles[p_i].inertia[1][1] = restart_data["inertia"][p_i];
+          particles[p_i].inertia[2][2] = restart_data["inertia"][p_i];
+
+          particles[p_i].pressure_location[0] = restart_data["pressure_x"][p_i];
+          particles[p_i].pressure_location[1] = restart_data["pressure_y"][p_i];
+          particles[p_i].pressure_location[2] = restart_data["pressure_z"][p_i];
+          particles[p_i].youngs_modulus = restart_data["youngs_modulus"][p_i];
+          particles[p_i].restitution_coefficient =
+            restart_data["restitution_coefficient"][p_i];
+          particles[p_i].friction_coefficient =
+            restart_data["friction_coefficient"][p_i];
+          particles[p_i].poisson_ratio = restart_data["poisson_ratio"][p_i];
+          particles[p_i].rolling_friction_coefficient =
+            restart_data["rolling_friction_coefficient"][p_i];
+          particles[p_i].initialise_last();
```

Comment:
what is this function? Why last and not end? you call this for all particles no?